docs/design-docs/04-arm-build.md

```diff
@@ -0,0 +1,66 @@
+# Title
+
+* Author(s): Mike Borozdin, @mikebz
+* Approver: Mengqi, Sunil
+
+## Why
+
+More people are getting Apple M1 machines.  The current docker images for
+functions do not work.  Here is an example:
+https://github.com/GoogleContainerTools/kpt/issues/2874
+While it's not a problem for CI/CD pipelines where architecture is mostly amd64
+for client development purposes the users are stuck.
+
+## Design
+
+Building arm64 and amd64 multi platform images is possible:
+https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/
+but it has it's caveats.  The build is done in a builder which needs to be 
+created:
+https://docs.docker.com/engine/reference/commandline/buildx_create/
+The caveat is that the multi platform images do not load into docker images:
+https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#-load-the-single-platform-build-result-to-docker-images---load
+
+Right now the build pipeline has several steps:
+1) build
+2) tag
+3) push
+
+In different scenarios only the first step is executed.  If we switch to 
+`buildx` we will need to build/tag/push in one step.  It is not clear if
+we need to build things locally (maybe the OWNERs can give guidance on that).
+
+However if we make the changes needed to the shell scripts that produce the
+images all that the users of docker images will need to do is just update
```

Comment:
Does it hurt the performance? When pulling a multi-arch image, is docker smart enough to only pull the matching part or everything for different arch? If the later, the image size to download is doubled or tripled (if we support arm32/v7 and arm64/v8)

Reply:
My understanding is that multi-arch images are all individual images that are all valid under a single tag. Your docker runtime knows what the local arch is, and it should only fetch the matching container. I can't think of a reason why this would cause a performance issue in practice.

Reply:
this one is a hard one to predict.  I am going to need to do a test probably to verify what this looks like.  We have a number of customers who are relying on basic images to be fairly speedy and their primary concern is the CI/CD pipeline.  One of the other things to do is to possibly add the emulation flag to the kpt binary...  this will basically allow people to emulate amd64 on a mac.

Reply:
Here are the findings:  the image is not 2x the size.  It looks like separate images are created for arm64 and amd64 and they are not bigger in size.  the top tag becomes a pointer.  Regular docker build and push results:
```
docker buildx imagetools inspect us-central1-docker.pkg.dev/mikebz-ex1/mbzcr/check-labels:v0.2
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
   "config": {
      "mediaType": "application/vnd.docker.container.image.v1+json",
      "size": 1865,
      "digest": "sha256:0013ea02a33011596b9ec69f8e2bf01de1d25c72dd790f7a4e6b34328c8dc38a"
   },
   "layers": [
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 2814512,
         "digest": "sha256:40e059520d199e1a1a259089077f2a0c879951c9a4540490bad3a0d7714c6ae7"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 3117030,
         "digest": "sha256:31e6c0326e6befcafbcc24f20fdf48cd199230f11ecc3f8cf9009f926f645163"
      }
   ]
}
```

Individual image looks like this:
<img width="677" alt="image" src="https://user-images.githubusercontent.com/260081/160758623-35658b74-93e0-4f00-b94c-c9ed32a5dd0d.png">



Then buildx result:
```
docker buildx imagetools inspect us-central1-docker.pkg.dev/mikebz-ex1/mbzcr/check-labels:v0.3                                  
Name:      us-central1-docker.pkg.dev/mikebz-ex1/mbzcr/check-labels:v0.3
MediaType: application/vnd.docker.distribution.manifest.list.v2+json
Digest:    sha256:15a41dced1a488bd2e0690fa047873931b673fe1b3a5a792e23b4984d1a71f4d
           
Manifests: 
  Name:      us-central1-docker.pkg.dev/mikebz-ex1/mbzcr/check-labels:v0.3@sha256:edea73063326e28693555bfd909c47c394af6719601235663c3203dfde43ae21
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/amd64
             
  Name:      us-central1-docker.pkg.dev/mikebz-ex1/mbzcr/check-labels:v0.3@sha256:56808f70266b9ba8a4c80df7059932313a1d41860828733de5e831ea3a35f562
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/arm64
  ```
The images are pushed simultaneously:
<img width="792" alt="image" src="https://user-images.githubusercontent.com/260081/160758468-b6c36fdb-aa11-4e71-ae36-f619c73a4b3b.png">

and the individual images look like this:

<img width="684" alt="image" src="https://user-images.githubusercontent.com/260081/160758537-50982908-a7e9-4feb-90d0-93f0bfdc386d.png">